### PR TITLE
JIT: Stop creating "location" commas

### PR DESCRIPTION
**Pull request by jakobbotsch**
Avoid creating commas that are on the LHS of an assignment or below an ADDR node.

These primarily were created by morph which has an `IND(COMMA(x, y)) -> COMMA(x, IND(y))` transformation that did not pay attention to whether it was on the left of an assignment.
The IR shape is pretty odd; the RHS of these commas are not actually evaluated in the traditional sense, but happen as part of the parent ASG, so the semantics of the construct ends up being confusing.
Additionally it complicates #83590.

I initially tried removing the `IND(COMMA(x, y)) -> COMMA(x, IND(y))` transformation entirely but this has a lot of diffs -- probably a bit of work to teach downstream phases about these new commas would be required. I gave up after doing that in a few separate places.

### CODE REVIEW

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -10048,21 +10046,25 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac, bool* optA
...
-            GenTree* propagatedThrow = fgPropagateCommaThrow(tree, op2->AsOp(), op1->gtFlags & GTF_ALL_EFFECT);
-            if (propagatedThrow != nullptr)
+            if ((op2 != nullptr) && fgIsCommaThrow(op2, true))
             {
-                return propagatedThrow;
+                GenTree* propagatedThrow = fgPropagateCommaThrow(tree, op2->AsOp(), op1->gtFlags & GTF_ALL_EFFECT);
+                if (propagatedThrow != nullptr)
+                {
+                    return propagatedThrow;
+                }
```
Comment (jakobbotsch) [inline src/coreclr/jit/morph.cpp:10067]: This transform was creating illegal IR like the following:
```scala
fgMorphTree BB01, STMT00017 (after)
               [000098] -ACXG+-----                         ▌  ASG       long  
               [000134] --CXG+-N---                         ├──▌  COMMA     long  
               [000137] --CXG+-----                         │  ├──▌  CALL help void   CORINFO_HELP_OVERFLOW
               [000136] -----+----- arg0 in rcx             │  │  └──▌  CNS_INT   int    0
               [000142] -----+-----                         │  └──▌  CNS_INT   long   0
               [000095] -----+-----                         └──▌  CNS_INT   long   0xFFFFFFFF
```

**Comment by jakobbotsch** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=215814&view=ms.vss-build-web.run-extensions-tab) seem to be due to new/missed CSE opportunities

The TP regressions are the same as in https://github.com/dotnet/runtime/pull/83590#issuecomment-1479256664. Apparently removing the call to `MorphBlock` from `PrepareDst` is enough to cause MSVC to change its inlining decisions for `fgMorphCopyBlock` and `fgMorphInitBlock`.

**Review by AndyAyersMS — APPROVED** [review]
These can still be created explicitly?

Wondering at what point we can transition from "avoid creating more of these" to "avoid having these all together" ?

File: src/coreclr/jit/morph.cpp [hunk 2]
```diff
@@ -11504,6 +11502,13 @@ GenTree* Compiler::fgPropagateCommaThrow(GenTree* parent, GenTreeOp* commaThrow,
     assert(fgGlobalMorph);
     assert(fgIsCommaThrow(commaThrow));
 
+    bool mightBeLocation = (parent->OperIsLocal() || parent->OperIsIndir()) && ((parent->gtFlags & GTF_DONT_CSE) != 0);
```
Comment (SingleAccretion) [inline src/coreclr/jit/morph.cpp:11505]: `parent->OperIsLocal` looks a bit unnecessary as `LCL_VAR/LCL_FLD` LHSs cannot have children, and `STORE_LCL_VAR/FLD` (when those start appearing here) won't need to be restricted like this.

File: src/coreclr/jit/morphblock.cpp [hunk 3]
```diff
@@ -1595,6 +1593,12 @@ GenTree* Compiler::fgMorphInitBlock(GenTree* tree)
 //
 GenTree* Compiler::fgMorphStoreDynBlock(GenTreeStoreDynBlk* tree)
 {
+    if (!tree->Data()->OperIs(GT_CNS_INT, GT_INIT_VAL))
+    {
+        // Data is a location.
+        tree->Data()->gtFlags |= GTF_DONT_CSE;
```
Comment (SingleAccretion) [inline src/coreclr/jit/morphblock.cpp:1599]: If we're relying on this NO_CSE, seems prudent to set it in import as well.

**Review by SingleAccretion — COMMENTED** [review]
There is also some code which can be deleted/simplified:

https://github.com/dotnet/runtime/blob/768c3ff106ddfd12e81f5dd456f30917e390cd8e/src/coreclr/jit/valuenum.cpp#L9842-L9844

https://github.com/dotnet/runtime/blob/768c3ff106ddfd12e81f5dd456f30917e390cd8e/src/coreclr/jit/optimizer.cpp#L8620

Nice to see the IR getting better \:).

File: src/coreclr/jit/morph.cpp (hunk 2 above)
Comment (jakobbotsch) [reply src/coreclr/jit/morph.cpp]: Removed that check.

File: src/coreclr/jit/morphblock.cpp (hunk 3 above)
Comment (jakobbotsch) [reply src/coreclr/jit/morphblock.cpp]: I decided to leave this here only, I think we should either set it on all locations in the importer or none of them, and today we don't set it on any of them. As you probably know we rely on this flag for the similar purpose in `fgMorphLocalVar` as well.

I guess today only the IND form of locations is allowed for `STORE_DYN_BLK`, but you could imagine if we allowed locations in general then something like `STORE_DYN_BLK(LCL_VAR int, LCL_VAR int, 4)` would have also been incorrectly handled without this.

**Review by SingleAccretion — APPROVED** [review]

**Comment by jakobbotsch** [comment]
Failures are known according to build analysis.

**Pull request merged**